docs/design/request-compression-trait.md

```diff
@@ -0,0 +1,546 @@
+Host: example.com
+Content-Type: application/json
+Content-Encoding: gzip
+
+// There was an encoding in header before compression
+POST /upload-data HTTP/1.1
+Host: example.com
+Content-Type: application/json
+Content-Encoding: br, gzip 
+```
+
+There is some logic behind the application of the [trait](https://smithy.io/2.0/spec/behavior-traits.html?highlight=requestcompression#requestcompression-trait), but it'll be handled by smithy so there's no need to worry about it. All that the SDK should care about is what type the http body will be once it gets to it. The http body could be any one of these four types:
+
+1. Source Content (streaming)
+
+2. Channel Content (streaming)
+
+3. Bytes (non-streaming)
+
+4. Empty (non-streaming)
+
+And all of those are handled differently. Here's how each one should be handled (in the same order as above):
+
+1. Compression will be applied for streaming HTTP bodies
+
+2. Compression will be applied for streaming HTTP bodies
+
+3. Compression will be applied for non streaming HTTP bodies
+
+4. No compression is applied
+
+Here is how it works in more detail:
+
+```kotlin
+val request = context.protocolRequest.toBuilder()
+
+request.body = when(request.body) {
+    is HttpBody.SourceContent -> // Apply compression for streaming HTTP bodies
+    is HttpBody.ChannelContent -> // Apply compression for streaming HTTP bodies
+    is HttpBody.Bytes -> // Apply compression for non streaming HTTP bodies
+    is HttpBody.Empty -> // Return as is (no compression is applied)
+    else -> throw ClientException("HttpBody type is not supported")
+}
+
+request.build()
+```
+
+## Compression Algorithms
+Given that Kotlin is multi-platform there should be multiple "implementations" of each compression algorithm. As of the time of writing this doc (11/07/2023) Gzip is the only supported compression algorithm by the server side and the only one we will be supporting for
+the SDK. For the time being, only Gzip for JVM will be supported specifically. This will be done by using the following packages from the Java standard library: `GZIPInputStream` and `GZIPOutputStream`. The other KMP targets are TODO's, and we have plans of using the CRT implementation for those.
+
+We'll be able to add new ones as they start becoming available/required using the compression algorithm interface we've written and
+by adding them to the list of supported algorithms. The same goes for non JVM KMP targets, these can be added by filling in the `actual` declarations for the target.
+
+#### Interfaces
+This interface will be used to represent a compression algorithm.
+
+```kotlin
+/**
+ * Represents a compression algorithm to be used for compressing request payloads on qualifying operations
+ */
+public interface CompressionAlgorithm {
+    /**
+     * The ID of the checksum algorithm
+     */
+    public fun algorithmId(): String
+
+    /**
+     * Compresses a non-stream payload
+     */
+    public fun compress(content: ByteArray?): ByteArray
```

Comment:
I'm not sure we need non-streaming vs streaming variants since both are covered by `ByteStream`. 

Do we want to just have the semantics of given an input `ByteStream` give me a compressed equivalent?